src/client/Client.cc

```diff
@@ -11941,12 +11944,12 @@ void Client::refresh_snapdir_attrs(Inode *in, Inode *diri) {
   in->uid = diri->uid;
   in->gid = diri->gid;
   in->nlink = 1;
-  in->mtime = diri->mtime;
-  in->ctime = diri->ctime;
+  in->mtime = diri->snaprealm->last_modified;
```

Comment:
From the kclient and Clent.cc code, it's still possible that for some `inodes` there could be no `cap` in some corner cases, that means the `diri->snaprealm` will be `NULL` ?

In this case IMO we should set the `mtime/ctime` to the global snaprealm's `last_modified/change_attr`.

Reply:
I check both the kclient and Client.cc code, when exporting caps it will remove the `cap` and then it will adjust the snaprealms of the inode. And then it maybe setting the `in->snaprealm` to `NULL`.

Reply:
I might be missing this - could you point me to the relevant function name in user-space client @lxbsz?

Reply:
Sure.

`void Client::handle_cap_export()` --> `remove_cap(&cap, false)`:

```
void Client::remove_cap(Cap *cap, bool queue_release)
{
 ...
  if (!in.is_any_caps()) {
    ldout(cct, 15) << __func__ << " last one, closing snaprealm " << in.snaprealm << dendl;
    in.snaprealm_item.remove_myself();
    put_snap_realm(in.snaprealm);
    in.snaprealm = 0;    <=============== here will set it to nullptr.
  }
}
```

Reply:
Ah! OK - this instance (the only one). Hmmm... This happens when the migrator is exporting a subtree (frozen?), isn't it?

Reply:
Yeah, I think so.

Reply:
In that case, does the nullptr case needs to be handled? I'm guessing the MDS does not send back a snap trace when a subtree in under migration.

Reply:
I am not really sure about this. But I have seen several times that the `in.snaprealm` is `nullptr` in both kernel and userspace clients. Maybe we at least should print some warnings when we see `in.snaprealm` is `nullptr` ?

Reply:
Yeh - that makes sense. For this change, I'll give put it through test once our testing infra is back functioning. If it crashes (due to snaprealm being null), it'll be a good learning exercise to figure when the MDS would send back a snap trace for an inode with no caps. A bit of reverse engineering :)

Reply:
Yeah, sound good to me.